sub-packages/bionemo-scdl/simple-benchmark/scdl_speedtest.py

```diff
@@ -0,0 +1,1315 @@
+    Returns:
+        torch.Tensor: A sparse CSR tensor with the same shape as the input.
+    """
+    nz = dense != 0
+    row, col = nz.nonzero(as_tuple=True)
+    vals = dense[nz]
+    counts = torch.bincount(row, minlength=dense.size(0))
+    crow = torch.cat([torch.tensor([0], dtype=torch.long), counts.cumsum(0)])
+    return torch.sparse_csr_tensor(crow, col.to(torch.long), vals, dense.shape)
+
+
+def collate_annloader_batch(batch):
+    """Collate function for AnnData dataset to match SCDL output format."""
+    return dense_to_csr(batch.X)
+
+
+@dataclass
+class BenchmarkResult:
+    """Results from benchmarking a dataloader."""
+
+    name: str
+    disk_size_mb: float
+    setup_time_seconds: float
+    warmup_time_seconds: float
+    total_iteration_time_seconds: float
+    average_batch_time_seconds: float
+    total_batches: int
+    total_samples: int
+    samples_per_second: float
+    batches_per_second: float
+    peak_memory_mb: float
+    average_memory_mb: float
+    gpu_memory_mb: float = 0.0
+    errors: List[str] = None
+
+    # Instantiation metrics
+    instantiation_time_seconds: Optional[float] = None
+    peak_memory_during_instantiation_mb: Optional[float] = None
+    memory_before_instantiation_mb: Optional[float] = None
+    memory_after_instantiation_mb: Optional[float] = None
+
+    # Configuration metadata
+    madvise_interval: Optional[int] = None
+    data_path: Optional[str] = None
+    max_time_seconds: Optional[float] = None
+    shuffle: Optional[bool] = None
+
+    # Conversion metrics
+    conversion_time_seconds: Optional[float] = None
+    conversion_performed: bool = False
+
+    # Load metrics (for AnnData baseline)
+    load_time_seconds: Optional[float] = None
+    load_performed: bool = False
+
+    # Warmup metrics
+    warmup_samples: int = 0
+    warmup_batches: int = 0
+
+    # Speed metrics
+    total_speed_samples_per_second: float = 0.0
+    post_warmup_speed_samples_per_second: float = 0.0
+
+    def __post_init__(self):
+        """Post-initialization hook for BenchmarkResult.
+
+        Ensures that the 'errors' attribute is always a list after initialization,
+        even if not provided. This prevents issues with mutable default arguments.
+        """
+        if self.errors is None:
+            self.errors = []
+
+    @classmethod
+    def from_raw_metrics(
+        cls,
+        name: str,
+        madvise_interval: Optional[int] = None,
+        data_path: Optional[str] = None,
+        max_time_seconds: Optional[float] = None,
+        shuffle: Optional[bool] = None,
+        total_samples: int = 0,
+        total_batches: int = 0,
+        setup_time: float = 0.0,
+        warmup_time: float = 0.0,
+        iteration_time: float = 0.0,
+        disk_size_mb: float = 0.0,
+        gpu_memory_mb: float = 0.0,
+        warmup_samples: int = 0,
+        warmup_batches: int = 0,
+        elapsed_time: float = 0.0,
+        instantiation_metrics: Optional[Dict[str, float]] = None,
```

Comment:
Maybe use a `field(defaultfactory=dict)` rather than `None`, unless you really want it to be optional.